File: Codility/04. Counting Elements/FrogRiverOne/Solution.swift
// https://app.codility.com/programmers/lessons/4-counting_elements/frog_river_one/
// Painless

public func solution(_ X : Int, _ A : inout [Int]) -> Int {
    var M = [Bool](repeating: false, count: X)

    var R = -1
    var S = 0
    
    var I = 0
    while I < A.count && R < 0 {
        let K = A[I] - 1
        if !M[K] {
            M[K] = true
            S += 1
        }

        if S == X {
            R = I
        }

        I += 1
    }

    return R
}

File: Codility/04. Counting Elements/FrogRiverOne/Solution.py
# https://app.codility.com/programmers/lessons/4-counting_elements/frog_river_one/
# Painless

from typing import List


def solution(X: int, A: List[int]) -> int:
    """Track the set of still-missing positions, returning the time it becomes empty."""
    # Time: O(n + X)   Space: O(X)
    needed = set(range(1, X + 1))

    for time, pos in enumerate(A):
        needed.discard(pos)
        if not needed:
            return time

    return -1


def test() -> None:
    assert solution(5, [1, 3, 1, 4, 2, 3, 5, 4]) == 6
    assert solution(1, [1]) == 0
    assert solution(2, [1, 1, 1]) == -1
